### PR TITLE
[12.0][ADD] hr_employee_holiday: review employee holidays via employee form

### PR DESCRIPTION
**Pull request by alexey-pelykh**
image

### CODE REVIEW

**Comment by alexey-pelykh** [comment]
@aheficent @nikul-serpentcs please review if you don't mind

**Review by AaronHForgeFlow — APPROVED** [review]
Functional tests on runbot. It works. :+1:

File: hr_employee_holiday/__manifest__.py
Comment (nikul-serpentcs) [inline hr_employee_holiday/__manifest__.py]: @alexey-pelykh no need to here, because of `hr_holidays` depends on `hr`

Comment (alexey-pelykh) [reply hr_employee_holiday/__manifest__.py]: I got your point, yet I'm not 100% sure about transient dependencies - sometimes other reviews mentioned that it's needed to include those. @pedrobaeza bring in some light please - since I had doubt even when wrote that line

Comment (AaronHForgeFlow) [reply hr_employee_holiday/__manifest__.py]: I think, in case no direct data from hr module is used (specific view, security group. etc) you can remove the dependency.

Comment (alexey-pelykh) [reply hr_employee_holiday/__manifest__.py]: Got it, done!

**Comment by alexey-pelykh** [comment]
@nikul-serpentcs changes made, please re-review

@sswapnesh since you're actively working with leaves in 12.0, this OCA module may be of interest for you

**Comment by sswapnesh** [comment]
Thanks for the ping @alexey-pelykh 
I will take a look once I am done with Odoo (At least Some Important Apps)
 Ps: I haven't tested any single OCA module yet

File: hr_employee_holiday/README.rst
Comment (sswapnesh) [inline hr_employee_holiday/README.rst:71]: IMO we should use https://github.com/OCA/hr/tree/12.0 (or https://github.com/OCA/hr/)  instead of full module link as we are referring to Project here.

File: hr_employee_holiday/__manifest__.py
Comment (sswapnesh) [inline hr_employee_holiday/__manifest__.py:15]: dot in the end.

File: hr_employee_holiday/views/hr_employee.xml
Comment (sswapnesh) [inline hr_employee_holiday/views/hr_employee.xml]: health ?

File: hr_employee_holiday/views/hr_employee.xml
Comment (sswapnesh) [inline hr_employee_holiday/views/hr_employee.xml]: Better to add decoration here (already there in Leaves tree view)
`<tree decoration-danger="state == 'refuse'" decoration-info="state == 'draft'" string="Allocation Requests" decoration-bf="message_needaction == True">`

File: hr_employee_holiday/views/hr_employee.xml
Comment (sswapnesh) [inline hr_employee_holiday/views/hr_employee.xml]: Better to add decoration here (already there in Leaves tree view)
`<tree decoration-danger="state == 'refuse'" decoration-info="state == 'draft'" string="Allocation Requests" decoration-bf="message_needaction == True">`

File: hr_employee_holiday/views/hr_employee.xml
Comment (sswapnesh) [inline hr_employee_holiday/views/hr_employee.xml:55]: I would suggest to use `request_date_from` and `request_date_to` here so it will be aligned with Form view as well.
Eg.
screenshot

Form view:
screenshot

File: hr_employee_holiday/views/hr_employee.xml
Comment (sswapnesh) [inline hr_employee_holiday/views/hr_employee.xml:53]: and also I would suggest you to pass `employee_id` in  `context` here so It displays leave counts based on employee.
https://github.com/odoo/odoo/blob/12.0/addons/hr_holidays/models/hr_leave_type.py#L203

screenshot

**Comment by sswapnesh** [comment]
This looks Interesting (Tested with Superuser also)
Link: http://3352230-537-362d28.runbot1.odoo-community.org/web#id=1&action=121&model=hr.employee&view_type=form&menu_id=90

Employee is not visible when we open form view from Requests

screenshot
screessssssssssssnshot

peek 2018-12-30 00-42


Not sure if this is because of https://github.com/odoo/odoo/blob/12.0/addons/hr_holidays/models/hr_leave.py#L496

File: hr_employee_holiday/README.rst
Comment (alexey-pelykh) [reply hr_employee_holiday/README.rst:71]: This file is generated to whatever needed :)

File: hr_employee_holiday/views/hr_employee.xml
Comment (alexey-pelykh) [reply hr_employee_holiday/views/hr_employee.xml]: Fixed

File: hr_employee_holiday/views/hr_employee.xml
Comment (alexey-pelykh) [reply hr_employee_holiday/views/hr_employee.xml]: good point

File: hr_employee_holiday/views/hr_employee.xml
Comment (alexey-pelykh) [reply hr_employee_holiday/views/hr_employee.xml:55]: Under the hood it's a datetime, not date (which is bad, IMO, this `request_*` nonsense), same as in _Leaves > Managers > Leaves > All_

Comment (sswapnesh) [reply hr_employee_holiday/views/hr_employee.xml:55]: Yes that's also True, But this was to prevent confusion When someone open leave from from here.

Comment (alexey-pelykh) [reply hr_employee_holiday/views/hr_employee.xml:55]: Well, this confusion originates from the Odoo hr_holidays module, and if behaviour will change there - this module will adopt new one as well

**Comment by sswapnesh** [comment]
@alexey-pelykh  Are you getting same issue (https://github.com/OCA/hr/pull/537#issuecomment-450514001) in you local instance to ?

**Comment by alexey-pelykh** [comment]
@sswapnesh looking into that

File: hr_employee_holiday/views/hr_employee.xml
Comment (alexey-pelykh) [reply hr_employee_holiday/views/hr_employee.xml:53]: done

File: hr_employee_holiday/views/hr_employee.xml
Comment (alexey-pelykh) [reply hr_employee_holiday/views/hr_employee.xml]: done

**Comment by alexey-pelykh** [comment]
@sswapnesh re: issue with empty employee on leave requests - same here, it's because https://github.com/odoo/odoo/blob/c72403d2ba79f0d5091183e33240034f2fdc6f73/addons/hr_holidays/views/hr_leave_views.xml#L234 I think, because seems that `hr_leave_view_form_manager` have to be used forcibly..

**Comment by alexey-pelykh** [comment]
@sswapnesh just pushed changes to overcome that (by specifying proper form ID explicitly)

**Comment by sswapnesh** [comment]
I will check this and get back to you. Thanks for update.

File: hr_employee_holiday/views/hr_employee.xml
Comment (nikul-serpentcs) [inline hr_employee_holiday/views/hr_employee.xml]: @alexey-pelykh Can you please improve string `string="New request..."`

Comment (alexey-pelykh) [reply hr_employee_holiday/views/hr_employee.xml]: Done

**Comment by sswapnesh** [comment]
@alexey-pelykh Works well now. :+1:

**Comment by alexey-pelykh** [comment]
@sswapnesh it would be great to get a "approve" review, if you don't mind :)

**Review by sswapnesh — APPROVED** [review]
Tested on Latest Runbot and seems working fine to me.

**Comment by sswapnesh** [comment]
@pedrobaeza  Seems like @OCA-git-bot counts all Approvals (IMO only Approval of maintainer/member should be counted)

**Comment by pedrobaeza** [comment]
Other reviewers approval counts, although there should be one PSC reviewer. It's the task of the person who presses the merge button to see if reviews are enough or not, and the people with this power are PSC, so no special problem. This bot activity is more a reminder for not letting PRs without merge.

**Comment by pedrobaeza** [comment]
On the module itself, this is the opposite paradigm that follows Odoo in its design: it avoids to bloat a form with submodels in the same view. That's why smart-buttons were created: for summarizing easily and with not too much performance cost a general status, and clicking it for going to detail.

Why existing smart-buttons don't serve you, Alexey?

I'm :-1: on this approach, but if enough people say it's good, OCA can host all the opinions. It's enough to not install these modules for me and the people that thinks the same :wink:

**Comment by alexey-pelykh** [comment]
@pedrobaeza I was using the `editable="bottom"` on that field, and the "Add line..." button (did I get it right which smart buttons do you refer to?) but this was not ok for leaves - due to "brilliant" wizard-alike-but-not-wizard-edit-form o the request and allocation, there entities were planned to _not_ be quick-created, thus I had to create a button that calls a `form`, and even specific one, because default resolution brings up wrong form.

**Comment by pedrobaeza** [comment]
From your snapshot, I mean these smart-buttons:

50423270-e3e51c80-085b-11e9-8737-c80734439186

If they are not enough, then let's expand them, but not add one2many fields inside the form.

**Comment by alexey-pelykh** [comment]
oh those smart buttons! as you can see, it seems that they never worked actually?

**Comment by pedrobaeza** [comment]
Then we should fix them. I think you have already put an issue about that.

**Comment by alexey-pelykh** [comment]
@pedrobaeza got your point. So, to summarise: there's a "Leaves Left" smart-button, that opens up this:
image
which is grouped by type (but without actions) and looks like
image

W/o this module there's no convenient way to create new allocations/request for employee within card of that employee and approve/decline from there - that's the only real difference.

The inability to approve-decline is explained by fact that smart button leads to "hr.leave.report" model

**Comment by pedrobaeza** [comment]
Then:

1. Add that buttons on this view or a similar more generic one, but not as one2many.
2. Add the needed smart-buttons, but again not as one2many.

Get my point?

**Comment by alexey-pelykh** [comment]
"one2many on form is bad" - I got your point :)

I don't think that adding "action buttons with states" to `hr.leave.report` would be a good idea either - since it's a report?

**Comment by pedrobaeza** [comment]
> "one2many on form is bad" - I got your point :)

It's bad for being added massively.

> I don't think that adding "action buttons with states" to hr.leave.report would be a good idea either - since it's a report?

You can granulate some smart-buttons for logical filters, but of course you mustn't put an smart-button for each of the possible combinations. Seeing your screenshots, there's only 2 smart-buttons (allocations an leaves), and custom tree views for them.

**Comment by alexey-pelykh** [comment]
Sorry, didn't get it about "added massively"?

And did I got you correctly that I should add allocations & leaves smart-button, that would leave to some sort of a proxy model with many2many computed relation?

**Comment by sswapnesh** [comment]
@alexey-pelykh  Do you have any solution to close this form after Save ?
screenshot
screenshot

**Comment by alexey-pelykh** [comment]
@sswapnesh that would deny ability to edit, as far as I can see

**Pull request closed without merging**
